File: android/app/src/main/java/com/termux/extensions/RunTargetRelay.kt
package com.termux.extensions

import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.Response
import okhttp3.WebSocket
import okhttp3.WebSocketListener
import okio.ByteString
import okio.ByteString.Companion.toByteString
import org.json.JSONObject
import java.net.BindException
import java.net.InetAddress
import java.net.InetSocketAddress
import java.net.ServerSocket
import java.net.Socket
import java.net.URI
import java.util.concurrent.ConcurrentHashMap
import java.util.concurrent.CountDownLatch
import java.util.concurrent.ExecutorService
import java.util.concurrent.Executors
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicBoolean
import java.util.concurrent.atomic.AtomicLong

internal data class AndroidRunTargetRoute(
    val ticket: String,
    val tunnelPath: String,
    val preferredPort: Int,
    val originalUrl: String,
    val label: String? = null,
)

internal data class AndroidRunTargetRouteSet(
    val ownerId: String,
    val shellId: String,
    val relayGroupId: String,
    val primary: AndroidRunTargetRoute,
    val additional: List<AndroidRunTargetRoute>,
)

internal fun normalizeAndroidRunTargetRoute(
    raw: JSONObject,
    requireLabel: Boolean = false,
): AndroidRunTargetRoute {
    val ticket = raw.optString("ticket").trim()
    val tunnelPath = raw.optString("tunnelPath").trim()
    val preferredPort = raw.optInt("preferredPort", -1)
    val originalUrl = raw.optString("originalUrl").trim()
    require(ticket.matches(Regex("^[0-9a-f]{64}$"))) { "Run target ticket is invalid" }
    require(tunnelPath == "/api/run-targets/$ticket/tunnel") {
        "Run target tunnel path is invalid"
    }
    require(preferredPort in 1..65535) { "Run target preferred port is invalid" }
    val uri = URI(originalUrl)
    require(uri.scheme.equals("http", ignoreCase = true) && uri.userInfo == null) {
        "Run target URL must be credential-free HTTP"
    }
    require(uri.host in setOf("127.0.0.1", "localhost", "::1")) {
        "Run target URL must address server loopback"
    }
    require((if (uri.port >= 0) uri.port else 80) == preferredPort) {
        "Run target URL port does not match preferredPort"
    }
    val label = raw.optString("label").trim().ifEmpty { null }
    require(!requireLabel || label != null) { "Run target auxiliary route label is required" }
    return AndroidRunTargetRoute(ticket, tunnelPath, preferredPort, originalUrl, label)
}

internal fun normalizeAndroidRunTargetRouteSet(raw: JSONObject): AndroidRunTargetRouteSet {
    if (!raw.has("primary")) {
        val primary = normalizeAndroidRunTargetRoute(raw)
        return AndroidRunTargetRouteSet("", "", primary.ticket, primary, emptyList())
    }
    val ownerId = raw.optString("ownerId").trim()
    val shellId = raw.optString("shellId").trim()
    require(ownerId.isNotEmpty()) { "Run target owner id is required" }
    require(shellId.isNotEmpty()) { "Run target shell id is required" }
    val relayGroupId = raw.optString("relayGroupId").trim()
    require(relayGroupId.matches(Regex("^[0-9a-f]{64}$"))) {
        "Run target relay group is invalid"
    }
    val primaryRaw = raw.optJSONObject("primary")
        ?: throw IllegalArgumentException("Run target primary route is required")
    val primary = normalizeAndroidRunTargetRoute(primaryRaw)
    val additionalRaw = raw.optJSONArray("additional")
        ?: throw IllegalArgumentException("Run target auxiliary routes are invalid")
    require(additionalRaw.length() <= 8) { "Run target supports at most 8 auxiliary routes" }
    val seenPorts = mutableSetOf(primary.preferredPort)
    val additional = buildList {
        for (index in 0 until additionalRaw.length()) {
            val routeRaw = additionalRaw.optJSONObject(index)
                ?: throw IllegalArgumentException("Run target auxiliary route is invalid")
            val route = normalizeAndroidRunTargetRoute(routeRaw, requireLabel = true)
            require(seenPorts.add(route.preferredPort)) {
                "Run target contains duplicate port ${route.preferredPort}"
            }
            add(route)
        }
    }
    require(relayGroupId == primary.ticket) { "Run target relay group is invalid" }
    return AndroidRunTargetRouteSet(ownerId, shellId, relayGroupId, primary, additional)
}

internal fun normalizeAndroidRunTargetProjection(raw: JSONObject): List<AndroidRunTargetRouteSet> {
    require(raw.optString("dto") == "RunTargetRouteProjection") {
        "Run target projection is invalid"
    }
    require(raw.optInt("version", -1) == 1) { "Run target projection version is invalid" }
    val groups = raw.optJSONArray("groups")
        ?: throw IllegalArgumentException("Run target projection groups are invalid")
    return buildList {
        for (index in 0 until groups.length()) {
            add(
                normalizeAndroidRunTargetRouteSet(
                    groups.optJSONObject(index)
                        ?: throw IllegalArgumentException("Run target projection group is invalid"),
                ),
            )
        }
    }
}

internal fun localAndroidRunTargetUrl(route: AndroidRunTargetRoute): String {
    val source = URI(route.originalUrl)
    return URI(
        "http",
        null,
        "127.0.0.1",
        route.preferredPort,
        source.rawPath,
        source.rawQuery,
        source.rawFragment,
    ).toASCIIString()
}

private fun tunnelWebSocketUrl(frameworkBaseUrl: String, tunnelPath: String): String {
    val base = URI(frameworkBaseUrl.trimEnd('/') + "/")
    val resolved = base.resolve(tunnelPath)
    val scheme = when (resolved.scheme.lowercase()) {
        "http" -> "ws"
        "https" -> "wss"
        else -> throw IllegalArgumentException(
            "Run target tunnel requires an HTTP or HTTPS framework origin",
        )
    }
    return URI(
        scheme,
        resolved.userInfo,
        resolved.host,
        resolved.port,
        resolved.rawPath,
        resolved.rawQuery,
        null,
    ).toASCIIString()
}

internal fun isAndroidConfiguredFrameworkLoopback(frameworkBaseUrl: String): Boolean {
    val host = runCatching {
        URI(frameworkBaseUrl.trimEnd('/') + "/").host
    }.getOrNull()?.removeSurrounding("[", "]")?.lowercase()
    return host in setOf("127.0.0.1", "localhost", "::1")
}

/** Owns native loopback listeners used to make remote Run Profile ports local. */
internal class RunTargetRelayManager(
    private val httpClient: OkHttpClient,
    private val executor: ExecutorService = Executors.newCachedThreadPool(),
    private val localityClassifier: (String) -> Boolean =
        ::isAndroidConfiguredFrameworkLoopback,
) {
    private data class GroupIdentity(
        val ownerId: String,
        val shellId: String,
    )

    private data class Entry(
        val groupId: String,
        val route: AndroidRunTargetRoute,
        val frameworkBaseUrl: String,
        val server: ServerSocket,
        val sockets: MutableSet<Socket> = ConcurrentHashMap.newKeySet(),
        val websockets: MutableSet<WebSocket> = ConcurrentHashMap.newKeySet(),
        val running: AtomicBoolean = AtomicBoolean(true),
        val acceptStopped: CountDownLatch = CountDownLatch(1),
    )

    private val lock = Any()
    private val mutationExecutor = Executors.newSingleThreadExecutor()
    private val entries = mutableMapOf<Int, Entry>()
    private val groupIdentities = mutableMapOf<String, GroupIdentity>()
    private var activeRoutesByShellId = emptyMap<String, AndroidRunTargetRouteSet>()
    private var hasAuthoritativeProjection = false
    private var projectionReady = false
    private var lastConfiguredLoopback: Boolean? = null
    private var lastError: String? = null
    private val projectionGeneration = AtomicLong(0)
    fun updateRouteProjection(
        rawProjection: JSONObject,
        frameworkBaseUrl: String,
        completion: (Result<Unit>) -> Unit = {},
    ) {
        val groups = normalizeAndroidRunTargetProjection(rawProjection)
        val next = groups.associateBy { it.shellId }
        val generation = projectionGeneration.incrementAndGet()
        synchronized(lock) {
            activeRoutesByShellId = next
            hasAuthoritativeProjection = true
            projectionReady = false
        }
        mutationExecutor.execute {
            val result = runCatching {
                val configuredLoopback = localityClassifier(frameworkBaseUrl)
                synchronized(lock) {
                    if (generation != projectionGeneration.get()) return@synchronized
                    require(hasAuthoritativeProjection) {
                        "Run target route authority is unavailable"
                    }
                    reconcileProjectionLocked(next, frameworkBaseUrl, configuredLoopback)
                    lastConfiguredLoopback = configuredLoopback
                    lastError = null
                    projectionReady = true
                }
            }
            result.exceptionOrNull()?.let { error ->
                synchronized(lock) {
                    if (generation == projectionGeneration.get()) {
                        lastError = error.message ?: error.javaClass.simpleName
                        projectionReady = false
                    }
                }
            }
            completion(result)
        }
    }

    fun suspendRouteProjection() {
        synchronized(lock) {
            hasAuthoritativeProjection = false
            projectionReady = false
            projectionGeneration.incrementAndGet()
        }
    }

    fun isProjectionReady(): Boolean = synchronized(lock) {
        hasAuthoritativeProjection && projectionReady
    }

    fun debugSnapshot(): JSONObject = synchronized(lock) {
        JSONObject().apply {
            put("authorityAvailable", hasAuthoritativeProjection)
            put("projectionReady", projectionReady)
            put("configuredLoopback", lastConfiguredLoopback ?: JSONObject.NULL)
            put("lastError", lastError ?: JSONObject.NULL)
            put("projectedGroups", activeRoutesByShellId.size)
            put("listenerPorts", org.json.JSONArray(entries.keys.sorted()))
        }
    }

    fun stopAll() {
        synchronized(lock) {
            hasAuthoritativeProjection = false
            projectionReady = false
            activeRoutesByShellId = emptyMap()
            projectionGeneration.incrementAndGet()
            groupIdentities.clear()
            entries.values.toList().forEach(::stopEntryLocked)
            entries.clear()
            lastConfiguredLoopback = null
            lastError = null
        }
    }

    fun close() {
        stopAll()
        mutationExecutor.shutdownNow()
        executor.shutdownNow()
    }

    private fun reconcileProjectionLocked(
        next: Map<String, AndroidRunTargetRouteSet>,
        frameworkBaseUrl: String,
        configuredLoopback: Boolean,
    ) {
        val staleGroupIds = groupIdentities.entries
            .filter { (groupId, identity) ->
                val active = next[identity.shellId]
                active == null ||
                    active.ownerId != identity.ownerId ||
                    active.relayGroupId != groupId ||
                    !groupMatchesLocked(active)
            }
            .map { it.key }
        staleGroupIds.forEach(::stopGroupLocked)
        if (configuredLoopback) {
            groupIdentities.keys.toList().forEach(::stopGroupLocked)
            return
        }
        next.values.forEach { routeSet ->
            startOrReuseGroupLocked(routeSet, frameworkBaseUrl)
        }
    }

    private fun startOrReuseGroupLocked(
        routeSet: AndroidRunTargetRouteSet,
        frameworkBaseUrl: String,
    ) {
        val routes = listOf(routeSet.primary) + routeSet.additional
        routes.forEach { route ->
            val occupied = entries[route.preferredPort]
            if (occupied != null && occupied.groupId != routeSet.relayGroupId) {
                val prefix = route.label?.let { "$it " }.orEmpty()
                throw IllegalStateException(
                    "${prefix}port ${route.preferredPort} is already in use",
                )
            }
        }
        if (groupMatchesLocked(routeSet)) return
        stopGroupLocked(routeSet.relayGroupId)
        try {
            routes.forEach { route ->
                try {
                    startEntryLocked(routeSet.relayGroupId, route, frameworkBaseUrl)
                } catch (error: BindException) {
                    val prefix = route.label?.let { "$it " }.orEmpty()
                    throw IllegalStateException(
                        "${prefix}port ${route.preferredPort} is already in use",
                        error,
                    )
                }
            }
            groupIdentities[routeSet.relayGroupId] = GroupIdentity(
                ownerId = routeSet.ownerId,
                shellId = routeSet.shellId,
            )
        } catch (error: Exception) {
            stopGroupLocked(routeSet.relayGroupId)
            throw error
        }
    }

    private fun acceptLoop(entry: Entry) {
        try {
            while (entry.running.get()) {
                val socket = try {
                    entry.server.accept()
                } catch (_: Exception) {
                    break
                }
                if (!entry.running.get()) {
                    socket.close()
                    break
                }
                entry.sockets.add(socket)
                executor.execute { bridge(entry, socket) }
            }
        } finally {
            entry.acceptStopped.countDown()
        }
    }

    private fun startEntryLocked(
        groupId: String,
        route: AndroidRunTargetRoute,
        frameworkBaseUrl: String,
    ) {
        val server = ServerSocket()
        // Exact Framework-Shell replacement must be able to reclaim the same
        // loopback port immediately after the prior owned listener closes.
        // SO_REUSEADDR does not permit a second simultaneous listener here;
        // an unrelated live owner still produces BindException.
        server.reuseAddress = true
        try {
            server.bind(
                InetSocketAddress(
                    InetAddress.getByName("127.0.0.1"),
                    route.preferredPort,
                ),
            )
        } catch (error: Exception) {
            server.close()
            throw error
        }
        val entry = Entry(groupId, route, frameworkBaseUrl, server)
        entries[route.preferredPort] = entry
        executor.execute { acceptLoop(entry) }
    }

    private fun groupMatchesLocked(routeSet: AndroidRunTargetRouteSet): Boolean {
        val expected = listOf(routeSet.primary) + routeSet.additional
        val current = entries.values.filter { it.groupId == routeSet.relayGroupId }
        return current.size == expected.size && expected.all { route ->
            entries[route.preferredPort]?.let { entry ->
                entry.groupId == routeSet.relayGroupId && entry.route.ticket == route.ticket
            } == true
        }
    }

    private fun stopGroupLocked(groupId: String) {
        groupIdentities.remove(groupId)
        entries.values.filter { it.groupId == groupId }.toList().forEach(::stopEntryLocked)
    }

    private fun bridge(entry: Entry, socket: Socket) {
        socket.tcpNoDelay = true
        val closed = AtomicBoolean(false)
        var websocket: WebSocket? = null

        fun closePair() {
            if (!closed.compareAndSet(false, true)) return
            entry.sockets.remove(socket)
            websocket?.let {
                entry.websockets.remove(it)
                it.cancel()
            }
            runCatching { socket.close() }
        }

        val listener = object : WebSocketListener() {
            override fun onOpen(webSocket: WebSocket, response: Response) {
                executor.execute {
                    val buffer = ByteArray(64 * 1024)
                    try {
                        val input = socket.getInputStream()
                        while (!closed.get()) {
                            val read = input.read(buffer)
                            if (read < 0) break
                            if (read > 0 && !webSocket.send(buffer.toByteString(0, read))) break
                        }
                    } catch (_: Exception) {
                    } finally {
                        closePair()
                    }
                }
            }

            override fun onMessage(webSocket: WebSocket, bytes: ByteString) {
                try {
                    synchronized(socket) {
                        socket.getOutputStream().apply {
                            write(bytes.toByteArray())
                            flush()
                        }
                    }
                } catch (_: Exception) {
                    closePair()
                }
            }

            override fun onMessage(webSocket: WebSocket, text: String) {
                closePair()
            }

            override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
                closePair()
            }

            override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
                closePair()
            }

            override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
                closePair()
            }
        }
        val request = Request.Builder()
            .url(tunnelWebSocketUrl(entry.frameworkBaseUrl, entry.route.tunnelPath))
            .build()
        val created = httpClient.newWebSocket(request, listener)
        websocket = created
        if (closed.get()) created.cancel() else entry.websockets.add(created)
    }

    private fun stopEntryLocked(entry: Entry) {
        if (entries[entry.route.preferredPort] === entry) {
            entries.remove(entry.route.preferredPort)
        }
        entry.running.set(false)
        runCatching { entry.server.close() }
        runCatching { entry.acceptStopped.await(1, TimeUnit.SECONDS) }
        entry.sockets.toList().forEach { runCatching { it.close() } }
        entry.sockets.clear()
        entry.websockets.toList().forEach { it.cancel() }
        entry.websockets.clear()
    }
}

File: desktop_client/electron/src/main/run-target-relay.ts
import * as net from "node:net";

import type {
  ElectronRunTargetAuxiliaryRoute,
  ElectronRunTargetRoute,
  ElectronRunTargetRouteSet,
} from "../shared/app-view-contracts";

const LOOPBACK_HOST = "127.0.0.1";
const TICKET_RE = /^[0-9a-f]{64}$/;
const MAX_AUXILIARY_PORTS = 8;
const MAX_WEBSOCKET_BACKLOG_BYTES = 4 * 1024 * 1024;

type RelayEntry = {
  groupId: string;
  route: ElectronRunTargetRoute;
  server: net.Server;
  sockets: Set<net.Socket>;
  websockets: Set<WebSocket>;
};

type GroupIdentity = {
  ownerId: string;
  shellId: string;
};

export type ElectronRunTargetRouteProjection = {
  dto: "RunTargetRouteProjection";
  version: 1;
  groups: ElectronRunTargetRouteSet[];
};

function errorCode(error: unknown): string {
  return error && typeof error === "object" && "code" in error
    ? String(error.code || "")
    : "";
}

function loopbackUrl(rawUrl: string, expectedPort: number): URL {
  const url = new URL(rawUrl);
  if (url.protocol !== "http:" || url.username || url.password) {
    throw new Error("Run target URL must be credential-free HTTP");
  }
  if (!["127.0.0.1", "localhost", "[::1]", "::1"].includes(url.hostname)) {
    throw new Error("Run target URL must address server loopback");
  }
  const effectivePort = Number(url.port || 80);
  if (effectivePort !== expectedPort) {
    throw new Error("Run target URL port does not match preferredPort");
  }
  return url;
}

export function normalizeRunTargetRoute(raw: unknown): ElectronRunTargetRoute {
  if (!raw || typeof raw !== "object" || Array.isArray(raw)) {
    throw new Error("Run target route is invalid");
  }
  const value = raw as Record<string, unknown>;
  const ticket = String(value.ticket || "").trim();
  const tunnelPath = String(value.tunnelPath || "").trim();
  const preferredPort = Number(value.preferredPort);
  const originalUrl = String(value.originalUrl || "").trim();
  if (!TICKET_RE.test(ticket)) throw new Error("Run target ticket is invalid");
  if (tunnelPath !== `/api/run-targets/${ticket}/tunnel`) {
    throw new Error("Run target tunnel path is invalid");
  }
  if (!Number.isInteger(preferredPort) || preferredPort < 1 || preferredPort > 65535) {
    throw new Error("Run target preferred port is invalid");
  }
  loopbackUrl(originalUrl, preferredPort);
  return {
    dto: "RunTargetRoute",
    version: 1,
    ticket,
    tunnelPath,
    preferredPort,
    originalUrl,
    expiresAt: Number(value.expiresAt) || 0,
  };
}

export function normalizeRunTargetRouteSet(raw: unknown): ElectronRunTargetRouteSet {
  if (!raw || typeof raw !== "object" || Array.isArray(raw)) {
    throw new Error("Run target route set is invalid");
  }
  const value = raw as Record<string, unknown>;
  if (!("primary" in value)) {
    const primary = normalizeRunTargetRoute(value);
    return {
      dto: "RunTargetRouteSet",
      version: 1,
      ownerId: "",
      shellId: "",
      relayGroupId: primary.ticket,
      primary,
      additional: [],
    };
  }
  const ownerId = String(value.ownerId || "").trim();
  const shellId = String(value.shellId || "").trim();
  if (!ownerId || !shellId) {
    throw new Error("Run target Framework-Shell ownership is invalid");
  }
  const relayGroupId = String(value.relayGroupId || "").trim();
  if (!TICKET_RE.test(relayGroupId)) {
    throw new Error("Run target relay group is invalid");
  }
  const primary = normalizeRunTargetRoute(value.primary);
  if (relayGroupId !== primary.ticket) {
    throw new Error("Run target relay group must identify the primary route");
  }
  if (!Array.isArray(value.additional) || value.additional.length > MAX_AUXILIARY_PORTS) {
    throw new Error("Run target auxiliary routes are invalid");
  }
  const seenPorts = new Set([primary.preferredPort]);
  const additional = value.additional.map((rawRoute): ElectronRunTargetAuxiliaryRoute => {
    const route = normalizeRunTargetRoute(rawRoute);
    const label = rawRoute && typeof rawRoute === "object" && !Array.isArray(rawRoute)
      ? String((rawRoute as Record<string, unknown>).label || "").trim()
      : "";
    if (!label) throw new Error("Run target auxiliary route label is required");
    if (seenPorts.has(route.preferredPort)) {
      throw new Error(`Run target contains duplicate port ${route.preferredPort}`);
    }
    seenPorts.add(route.preferredPort);
    return { ...route, label };
  });
  return {
    dto: "RunTargetRouteSet",
    version: 1,
    ownerId,
    shellId,
    relayGroupId,
    primary,
    additional,
  };
}

export function normalizeRunTargetRouteProjection(
  raw: unknown,
): ElectronRunTargetRouteProjection {
  if (!raw || typeof raw !== "object" || Array.isArray(raw)) {
    throw new Error("Run target route projection is invalid");
  }
  const value = raw as Record<string, unknown>;
  if (value.dto !== "RunTargetRouteProjection" || value.version !== 1) {
    throw new Error("Run target route projection is invalid");
  }
  if (!Array.isArray(value.groups)) {
    throw new Error("Run target route projection groups are invalid");
  }
  return {
    dto: "RunTargetRouteProjection",
    version: 1,
    groups: value.groups.map(normalizeRunTargetRouteSet),
  };
}

export function isConfiguredFrameworkLoopback(configuredFrameworkOrigin: string): boolean {
  const hostname = new URL(configuredFrameworkOrigin).hostname.toLowerCase();
  return ["127.0.0.1", "localhost", "[::1]", "::1"].includes(hostname);
}

export function localRunTargetUrl(route: ElectronRunTargetRoute): string {
  const url = loopbackUrl(route.originalUrl, route.preferredPort);
  url.hostname = LOOPBACK_HOST;
  url.port = String(route.preferredPort);
  return url.href;
}

function tunnelWebSocketUrl(configuredFrameworkOrigin: string, tunnelPath: string): string {
  const url = new URL(tunnelPath, configuredFrameworkOrigin);
  if (url.protocol === "http:") url.protocol = "ws:";
  else if (url.protocol === "https:") url.protocol = "wss:";
  else throw new Error("Run target tunnel requires an HTTP or HTTPS framework origin");
  return url.href;
}

function listen(server: net.Server, port: number): Promise<void> {
  return new Promise((resolvePromise, reject) => {
    const onError = (error: Error): void => reject(error);
    server.once("error", onError);
    server.listen({ host: LOOPBACK_HOST, port, exclusive: true }, () => {
      server.off("error", onError);
      resolvePromise();
    });
  });
}

function closeServer(server: net.Server): Promise<void> {
  if (!server.listening) return Promise.resolve();
  return new Promise((resolvePromise) => server.close(() => resolvePromise()));
}

export class RunTargetRelayManager {
  readonly #entries = new Map<number, RelayEntry>();
  readonly #groupIdentities = new Map<string, GroupIdentity>();
  #activeRoutesByShellId = new Map<string, ElectronRunTargetRouteSet>();
  #hasAuthoritativeProjection = false;
  #projectionReady = false;
  #projectionGeneration = 0;
  #lastConfiguredLoopback: boolean | null = null;
  #lastError: string | null = null;
  readonly #readyWaiters = new Set<{
    resolve: () => void;
    reject: (error: Error) => void;
    timeout: ReturnType<typeof setTimeout>;
  }>();
  #mutationQueue: Promise<void> = Promise.resolve();

  constructor(
    private readonly frameworkOrigin: () => string,
    private readonly localityClassifier:
      (frameworkOrigin: string) => boolean | Promise<boolean> =
        isConfiguredFrameworkLoopback,
  ) {}

  updateRouteProjection(rawProjection: unknown): Promise<void> {
    const projection = normalizeRunTargetRouteProjection(rawProjection);
    const next = new Map(projection.groups.map((group) => [group.shellId, group]));
    const generation = ++this.#projectionGeneration;
    this.#activeRoutesByShellId = next;
    this.#hasAuthoritativeProjection = true;
    this.#projectionReady = false;
    const operation = this.#mutationQueue.then(
      () => this.#reconcileProjection(next, generation),
    );
    this.#mutationQueue = operation.then(() => undefined, () => undefined);
    void operation.then(
      () => {
        if (generation !== this.#projectionGeneration) return;
        this.#projectionReady = true;
        this.#lastError = null;
        this.#resolveReadyWaiters();
      },
      (error: unknown) => {
        if (generation !== this.#projectionGeneration) return;
        const normalized = error instanceof Error ? error : new Error(String(error));
        this.#projectionReady = false;
        this.#lastError = normalized.message;
        this.#rejectReadyWaiters(normalized);
      },
    );
    return operation;
  }

  suspendRouteProjection(): void {
    this.#hasAuthoritativeProjection = false;
    this.#projectionReady = false;
    this.#projectionGeneration += 1;
  }

  waitUntilProjectionReady(timeoutMs = 10_000): Promise<void> {
    if (this.#hasAuthoritativeProjection && this.#projectionReady) {
      return Promise.resolve();
    }
    return new Promise((resolve, reject) => {
      const waiter = {
        resolve,
        reject,
        timeout: setTimeout(() => {
          this.#readyWaiters.delete(waiter);
          reject(new Error("Run target route authority did not become ready"));
        }, timeoutMs),
      };
      this.#readyWaiters.add(waiter);
    });
  }

  debugSnapshot(): Record<string, unknown> {
    return {
      authorityAvailable: this.#hasAuthoritativeProjection,
      projectionReady: this.#projectionReady,
      configuredLoopback: this.#lastConfiguredLoopback,
      lastError: this.#lastError,
      projectedGroups: this.#activeRoutesByShellId.size,
      listenerPorts: [...this.#entries.keys()].sort((left, right) => left - right),
    };
  }

  async #reconcileProjection(
    next: Map<string, ElectronRunTargetRouteSet>,
    generation: number,
  ): Promise<void> {
    const configuredLoopback = await this.localityClassifier(this.frameworkOrigin());
    if (generation !== this.#projectionGeneration) return;
    const staleGroupIds = [...this.#groupIdentities]
      .filter(([groupId, identity]) => {
        const active = next.get(identity.shellId);
        return !active ||
          active.ownerId !== identity.ownerId ||
          active.relayGroupId !== groupId ||
          !this.#groupMatches(active);
      })
      .map(([groupId]) => groupId);
    await Promise.all(staleGroupIds.map((groupId) => this.#stopGroup(groupId)));
    this.#lastConfiguredLoopback = configuredLoopback;
    if (configuredLoopback) {
      await Promise.all([...this.#groupIdentities.keys()].map((groupId) => (
        this.#stopGroup(groupId)
      )));
      return;
    }
    for (const routeSet of next.values()) {
      if (generation !== this.#projectionGeneration) return;
      await this.#startOrReuseGroup(routeSet);
    }
  }

  async #startOrReuseGroup(routeSet: ElectronRunTargetRouteSet): Promise<void> {
    const oldGroupIds = [...this.#groupIdentities]
      .filter(([, identity]) => (
        identity.ownerId === routeSet.ownerId && identity.shellId !== routeSet.shellId
      ))
      .map(([groupId]) => groupId);
    await Promise.all(oldGroupIds.map((groupId) => this.#stopGroup(groupId)));
    const existingPrimary = this.#entries.get(routeSet.primary.preferredPort);
    if (existingPrimary && existingPrimary.groupId !== routeSet.relayGroupId) {
      throw new Error(
        `Run target port ${routeSet.primary.preferredPort} is owned by another profile`,
      );
    }
    if (this.#groupMatches(routeSet)) {
      return;
    }
    await this.#stopGroup(routeSet.relayGroupId);

    const primary = this.#createEntry(routeSet.relayGroupId, routeSet.primary);
    try {
      await this.#startEntry(primary);
    } catch (error) {
      primary.server.close();
      if (errorCode(error) === "EADDRINUSE") {
        throw new Error(`Run target port ${routeSet.primary.preferredPort} is already in use`);
      }
      throw error;
    }

    try {
      for (const route of routeSet.additional) {
        const existing = this.#entries.get(route.preferredPort);
        if (existing && existing.groupId !== routeSet.relayGroupId) {
          throw new Error(
            `${route.label} port ${route.preferredPort} is owned by another run profile`,
          );
        }
        const entry = this.#createEntry(routeSet.relayGroupId, route);
        try {
          await this.#startEntry(entry);
        } catch (error) {
          entry.server.close();
          if (errorCode(error) === "EADDRINUSE") {
            throw new Error(`${route.label} port ${route.preferredPort} is already in use`);
          }
          throw error;
        }
      }
    } catch (error) {
      await this.#stopGroup(routeSet.relayGroupId);
      throw error;
    }
    this.#groupIdentities.set(routeSet.relayGroupId, {
      ownerId: routeSet.ownerId,
      shellId: routeSet.shellId,
    });
  }

  stopAll(): Promise<void> {
    const operation = this.#mutationQueue.then(() => this.#stopAllNow());
    this.#mutationQueue = operation.then(() => undefined, () => undefined);
    return operation;
  }

  #createEntry(groupId: string, route: ElectronRunTargetRoute): RelayEntry {
    const entry: RelayEntry = {
      groupId,
      route,
      server: net.createServer(),
      sockets: new Set(),
      websockets: new Set(),
    };
    entry.server.on("connection", (socket) => this.#accept(entry, socket));
    return entry;
  }

  async #startEntry(entry: RelayEntry): Promise<void> {
    await listen(entry.server, entry.route.preferredPort);
    entry.server.unref();
    this.#entries.set(entry.route.preferredPort, entry);
    entry.server.on("error", (error) => {
      console.error(`[te2-run-target] listener failed: ${error.message}`);
      void this.#stopGroup(entry.groupId);
    });
  }

  #groupMatches(routeSet: ElectronRunTargetRouteSet): boolean {
    const expected = [routeSet.primary, ...routeSet.additional];
    const current = [...this.#entries.values()].filter(
      (entry) => entry.groupId === routeSet.relayGroupId,
    );
    return current.length === expected.length && expected.every((route) => {
      const entry = this.#entries.get(route.preferredPort);
      return entry?.groupId === routeSet.relayGroupId && entry.route.ticket === route.ticket;
    });
  }

  async #stopAllNow(): Promise<void> {
    this.#hasAuthoritativeProjection = false;
    this.#projectionReady = false;
    this.#projectionGeneration += 1;
    this.#activeRoutesByShellId.clear();
    this.#groupIdentities.clear();
    this.#lastConfiguredLoopback = null;
    this.#lastError = null;
    await Promise.all([...this.#entries.values()].map((entry) => this.#stopEntry(entry)));
  }

  async #stopGroup(groupId: string): Promise<void> {
    this.#groupIdentities.delete(groupId);
    const entries = [...this.#entries.values()].filter((entry) => entry.groupId === groupId);
    await Promise.all(entries.map((entry) => this.#stopEntry(entry)));
  }

  #resolveReadyWaiters(): void {
    for (const waiter of this.#readyWaiters) {
      clearTimeout(waiter.timeout);
      waiter.resolve();
    }
    this.#readyWaiters.clear();
  }

  #rejectReadyWaiters(error: Error): void {
    for (const waiter of this.#readyWaiters) {
      clearTimeout(waiter.timeout);
      waiter.reject(error);
    }
    this.#readyWaiters.clear();
  }

  #accept(entry: RelayEntry, socket: net.Socket): void {
    if (this.#entries.get(entry.route.preferredPort) !== entry) {
      socket.destroy();
      return;
    }
    entry.sockets.add(socket);
    socket.pause();
    socket.setNoDelay(true);
    socket.unref();

    let closed = false;
    let opened = false;
    const websocket = new WebSocket(
      tunnelWebSocketUrl(this.frameworkOrigin(), entry.route.tunnelPath),
    );
    websocket.binaryType = "arraybuffer";
    entry.websockets.add(websocket);

    const closePair = (): void => {
      if (closed) return;
      closed = true;
      entry.sockets.delete(socket);
      entry.websockets.delete(websocket);
      if (!socket.destroyed) socket.destroy();
      if (
        websocket.readyState === WebSocket.CONNECTING ||
        websocket.readyState === WebSocket.OPEN
      ) {
        websocket.close();
      }
    };

    websocket.addEventListener("open", () => {
      if (closed) return;
      opened = true;
      socket.resume();
    });
    websocket.addEventListener("message", (event) => {
      if (closed || typeof event.data === "string" || !(event.data instanceof ArrayBuffer)) {
        closePair();
        return;
      }
      socket.write(Buffer.from(event.data), (error) => {
        if (error) closePair();
      });
    });
    websocket.addEventListener("close", closePair);
    websocket.addEventListener("error", closePair);
    socket.on("data", (chunk) => {
      if (!opened || websocket.readyState !== WebSocket.OPEN) {
        closePair();
        return;
      }
      if (websocket.bufferedAmount > MAX_WEBSOCKET_BACKLOG_BYTES) {
        closePair();
        return;
      }
      websocket.send(chunk);
    });
    socket.once("end", closePair);
    socket.once("close", closePair);
    socket.once("error", closePair);
  }

  async #stopEntry(entry: RelayEntry): Promise<void> {
    if (this.#entries.get(entry.route.preferredPort) === entry) {
      this.#entries.delete(entry.route.preferredPort);
    }
    for (const socket of entry.sockets) {
      if (!socket.destroyed) socket.destroy();
    }
    entry.sockets.clear();
    for (const websocket of entry.websockets) {
      if (
        websocket.readyState === WebSocket.CONNECTING ||
        websocket.readyState === WebSocket.OPEN
      ) {
        websocket.close();
      }
    }
    entry.websockets.clear();
    await closeServer(entry.server);
  }
}
